v2/car_reader_seeker.go

```diff
@@ -0,0 +1,107 @@
+package car
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	"golang.org/x/xerrors"
+)
+
+// CarReaderSeeker wraps CarOffsetWriter with a ReadSeeker implementation.
+// Note that Read and Seek are not thread-safe, they must not be called
+// concurrently.
+type CarReaderSeeker struct {
+	parentCtx context.Context
+	size      uint64
+	offset    int64
+
+	cow           *CarOffsetWriter // 🐮
+	reader        *io.PipeReader
+	writeCancel   context.CancelFunc
+	writeComplete chan struct{}
+}
+
+var _ io.ReadSeeker = (*CarReaderSeeker)(nil)
+
+func NewCarReaderSeeker(ctx context.Context, payloadCid cid.Cid, bstore blockstore.Blockstore, size uint64) *CarReaderSeeker {
+	cow := NewCarOffsetWriter(payloadCid, bstore)
+
+	return &CarReaderSeeker{
+		parentCtx:     ctx,
+		size:          size,
+		cow:           cow,
+		writeComplete: make(chan struct{}, 1),
+	}
+}
+
+// Note: not threadsafe
+func (c *CarReaderSeeker) Read(p []byte) (n int, err error) {
+	if uint64(c.offset) >= c.size {
+		return 0, fmt.Errorf("cannot read from offset %d >= file size %d", c.offset, c.size)
+	}
+
+	// Check if there's already a write in progress
+	if c.reader == nil {
+		// No write in progress, start a new write from the current offset
+		// in a go routine
+		writeCtx, writeCancel := context.WithCancel(c.parentCtx)
+		c.writeCancel = writeCancel
+		pr, pw := io.Pipe()
+		c.reader = pr
+
+		go func() {
+			err := c.cow.Write(writeCtx, pw, uint64(c.offset))
+			if err != nil && !xerrors.Is(err, context.Canceled) {
+				pw.CloseWithError(err)
+			} else {
+				pw.Close()
+			}
+			c.writeComplete <- struct{}{}
+		}()
+	}
+
+	return c.reader.Read(p)
+}
+
+// Note: not threadsafe
+func (c *CarReaderSeeker) Seek(offset int64, whence int) (int64, error) {
+	// Update the offset
+	switch whence {
+	case io.SeekStart:
+		if offset < 0 {
```

Comment:
We do that in Read() because ["Seeking to any positive offset is legal, but the behavior of subsequent I/O operations on the underlying object is implementation-dependent"](https://pkg.go.dev/io#Seeker).
As an example, http.ServeContent() seeks to the end of the file to get the file size.

Reply:
TIL.